pyramid/wr_file.go

```diff
@@ -12,28 +12,56 @@ type WRFile struct {
 
 	persisted bool
 	store     func(string) error
+	abort     func() error
+	aborted   bool
 }
 
-var errAlreadyPersisted = errors.New("file is already persisted")
+var (
+	errFilePersisted = errors.New("file is persisted")
+	errFileAborted   = errors.New("file is aborted")
+)
 
 // Store copies the closed file to all tiers of the pyramid.
 func (f *WRFile) Store(filename string) error {
+	if f.aborted {
+		return errFileAborted
+	}
+	if f.persisted {
+		return errFilePersisted
+	}
+	f.persisted = true
+
 	if err := validateFilename(filename); err != nil {
 		return err
 	}
 
-	err := f.File.Close()
-	if err != nil && !errors.Is(err, os.ErrClosed) {
-		return fmt.Errorf("closing file: %w", err)
+	if err := f.idempotentClose(); err != nil {
+		return err
 	}
 
+	return f.store(filename)
+}
+
+// Abort delete the file and cleans all traces of it.
+// If file was already stored, returns an error.
+func (f *WRFile) Abort() error {
 	if f.persisted {
-		return errAlreadyPersisted
+		return errFilePersisted
 	}
+	f.aborted = true
 
-	err = f.store(filename)
-	if err == nil {
-		f.persisted = true
+	if err := f.idempotentClose(); err != nil {
+		return err
+	}
+
+	return f.abort()
+}
+
+// idempotentClose is like Close(), but doesn't fail when the file is already closed.
+func (f *WRFile) idempotentClose() error {
```

Comment:
why not implement Close and check 'closed' bool as we did for abort?

Reply:
Since `*os.File` is embedded. User will still be able to write `weFile.File.Close()` without passing thru pyramid close. I can however not embed the os.File and implement a pass through for everything, but I rather avoid it.